icu4c/source/test/intltest/ucdtest.cpp

```diff
@@ -1024,3 +1027,107 @@ void UnicodeTest::TestIDCompatMath() {
+    char *unidataPath = getUnidataPath(path);
+    if(unidataPath == nullptr) {
+        fprintf(stderr, "TestBinaryPropertyUsingPpucd: unable to open ppucd.txt from ICU source tree\n");
+        errln("exiting early because unable to open ppucd.txt from ICU source tree");
+    }
+    char *unidataPathEnd = strchr(path, 0);
+    strcpy(unidataPathEnd, "ppucd.txt");
+    CharString ppucdPath(path, errorCode);
+    PreparsedUCD ppucd(ppucdPath.data(), errorCode);
+    if(errorCode.isFailure()) {
+        fprintf(stderr, "TestBinaryPropertyUsingPpucd: unable to open %s - %s\n",
+                ppucdPath.data(), errorCode.errorName());
+        errln("unable to open %s - %s\n",
+                ppucdPath.data(), errorCode.errorName());
+    }
+    BuiltInPropertyNames builtInPropNames;
+    ppucd.setPropertyNames(&builtInPropNames);
+
+    // Define which binary properties we want to compare
+    UProperty propsUnderTest[] = {
+        UCHAR_ID_COMPAT_MATH_START,
+        UCHAR_ID_COMPAT_MATH_CONTINUE,
+    };
+    uint32_t numProps = 2;
+
+    // Allocate & initialize UnicodeSets per binary property from PPUCD data
+    UnicodeSet *ppucdPropSets = new UnicodeSet[numProps];
+
+    // Iterate through PPUCD file, accumulating each line's data into each UnicodeSet per property
+    PreparsedUCD::LineType lineType;
+    UnicodeSet newValues;
+    while((lineType=ppucd.readLine(errorCode))!=PreparsedUCD::NO_LINE) {
+        if(ppucd.lineHasPropertyValues()) {
+            const UniProps *lineProps=ppucd.getProps(newValues, errorCode);
+
+            for(uint32_t i = 0; i < numProps; i++) {
+                UProperty prop = propsUnderTest[i];
+                int32_t propEnumVal = (int) prop;
+                // TODO: add or remove ranges according to value
+                if (lineProps->binProps[(propEnumVal)]) {
+                    ppucdPropSets[i].add(lineProps->start, lineProps->end);
+                } else {
+                    ppucdPropSets[i].remove(lineProps->start, lineProps->end);
+                }
+            }
+        }
+        if(errorCode.isFailure()) {
+            break;
+        }
+    }
+
+    if(errorCode.isFailure()) {
+        fprintf(stderr,
+                "genprops: error parsing or setting values from ppucd.txt"
+                " line %ld - %s\n",
+                (long)ppucd.getLineNumber(), errorCode.errorName());
+        errln("exiting early due to parsing error");
+    }
+
+    // Allocate & initialize UnicodeSets per binary property from ICU data
+    UnicodeSet *icuPropSets = new UnicodeSet[numProps];
+    for(uint32_t i = 0; i < numProps; i++) {
+        icuPropSets[i].applyIntPropertyValue(propsUnderTest[i], 1, errorCode);
+    }
+
+    // Assert that the PPUCD data and the ICU data are equivalent for all properties
+    for(uint32_t i = 0; i < numProps; i++) {
+        char msg[50];
+        snprintf(msg, 50, "ICU & PPUCD versions of property (enum = %d)",
+            (int) propsUnderTest[i]);
```

Comment:
Please look for a C++ function to format into a std::string and then use its .c_str() .

You can use `u_getPropertyName(prop, U_LONG_PROPERTY_NAME)` for more readable output than the integer value.

Reply:
Done. The `std::format` function is only available in C++20, so I used concatenation via `operator+` on `std::string`. I did what they do in Java, where if the first operand of what you're concatenating using `+` isn't already a `String`, then you start off with `""`. For `std::string`, the `operator+` won't work when the LHS isn't `std::string` (similar to Java).